File: botv2.py
import cv2
import pytesseract
import pyautogui

import requests
import numpy as np
from pynput.mouse import Controller

import undetected_chromedriver as uc
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
import time

# Path to the external text file containing account credentials
credentials_file = "credentials.txt"

# Instantiate the Chrome WebDriver
driver = webdriver.Chrome()
# Add any desired options to the options object, if needed
# options = uc.ChromeOptions()
# driver = uc.Chrome(options=options)

# URL to navigate to after logging in
target_url = "https://www.tiktok.com/@nikxphreaker/video/6993669592551836930"

# Path to the external text file containing comment texts
comments_file = "comments.txt"

# Read account credentials from the text file
accounts = []
with open(credentials_file, "r") as file:
    for line in file:
        username, password = line.strip().split(",")
        account = {"username": username, "password": password}
        accounts.append(account)

# Read comment texts from the text file
comments = []
with open(comments_file, "r") as file:
    for line in file:
        comment = line.strip()
        comments.append(comment)

# Iterate over each account
for account in accounts:
    # Open TikTok website
    driver.get("https://www.tiktok.com/login/phone-or-email/email")

    # Wait for the page to load
    time.sleep(2)

    # Find the login button and click it
    # login_button = driver.find_element(By.LINK_TEXT, "Log in")
    # login_button.click()

    # Wait for the login page to load
    time.sleep(2)

    # Find the username input field and enter the username
    username_input = driver.find_element(By.CSS_SELECTOR, "input[name='username']")
    username_input.send_keys(account["username"])
    time.sleep(5)

    # Find the password input field and enter the password
    password_input = driver.find_element(By.CSS_SELECTOR, "input[type='password']")
    password_input.send_keys(account["password"])
    time.sleep(5)

    # Press Enter to submit the login form
    password_input.send_keys(Keys.ENTER)

    #solving captcha
    time.sleep(10)

    #get element img captcha
    image_element = driver.find_element(By.CSS_SELECTOR, "img[id='captcha-verify-image']")
    image_url = image_element.get_attribute("src")
    print(image_url)

    # Fetch the image data from the URL
    response = requests.get(image_url)
    image_data = response.content

    # Convert the image data to a NumPy array
    image_array = np.frombuffer(image_data, np.uint8)

    # Read the image using OpenCV's imread function
    image = cv2.imdecode(image_array, cv2.IMREAD_COLOR)

    # Save the image with a specific filename
    filename = "output.jpg"  # Specify the desired filename
    cv2.imwrite(filename, image)

    # Load the captcha image
    captcha_image = cv2.imread('output.jpg', cv2.IMREAD_GRAYSCALE)

    # Apply image preprocessing
    _, thresholded_image = cv2.threshold(captcha_image, 127, 255, cv2.THRESH_BINARY)
    contours, _ = cv2.findContours(thresholded_image, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

    # Define a template shape for matching
    template = np.array([[0, 0], [0, 20], [20, 20], [20, 0]], dtype=np.int32)

    # Iterate over the contours and perform shape matching
    for contour in contours:
        epsilon = 0.02 * cv2.arcLength(contour, True)
        approx = cv2.approxPolyDP(contour, epsilon, True)

        # Perform shape matching
        match = cv2.matchShapes(template, approx, cv2.CONTOURS_MATCH_I3, 0)

        # If the match is below a certain threshold, consider it a match
        if match < 0.1:
            # Get the centroid of the contour
            M = cv2.moments(contour)
            cx = int(M['m10'] / M['m00'])
            cy = int(M['m01'] / M['m00'])

            # Simulate touch on the centroid
            pyautogui.moveTo(cx, cy)
            pyautogui.click()
            print(str(cx)+' '+str(cy))
            break  # Exit the loop after the first match

    # Wait for the login process to complete
    time.sleep(60)


    # Navigate to the target URL
    # driver.get(target_url)

    # Interact with TikTok content
    #for _ in range(5):
    
    # # Add likes to the post
    # like_button = driver.find_element(By.CSS_SELECTOR, "span[data-e2e='like-icon']")
    # like_button.click()
    # time.sleep(1)
    # # like_button = driver.find_element(By.CSS_SELECTOR, "button.like-button")
    # # like_button.click()
    # # time.sleep(1)

    # # Add comment to the post
    # # Select a random comment from the list
    # comment_text = random.choice(comments)
    # comment_input = driver.find_element(By.CSS_SELECTOR, "textarea.comment-input")
    # comment_input.send_keys(comment_text)
    # comment_input.send_keys(Keys.ENTER)
    # time.sleep(2)
        
    # Find a video element and extract the video URL
    #video_element = driver.find_element(By.CSS_SELECTOR, "a[href^='/v/']")
    #video_url = video_element.get_attribute("href")
    #print("Video URL:", video_url)

    # Move to the next video
    # driver.find_element_by_tag_name("body").send_keys(Keys.ARROW_RIGHT)

    # Logout and clear cookies for the next account
    driver.delete_all_cookies()

# Close the browser
driver.quit()

File: botv2.js
const webdriver = require('selenium-webdriver');
const { Builder, By, Key, until } = webdriver;
const chrome = require('selenium-webdriver/chrome');
const CaptchaSolver = require('tiktok-captcha-solver');
const puppeteer = require('puppeteer');

const logged_in_accounts = new Set(); // Keep track of logged-in accounts

const ua = require('random-useragent');
const user_agent = ua.getRandom();

async function loginTikTok(accounts, proxies, targetUrl, comments) {
  // Iterate over each account
  for (const account of accounts) {
    const browser = await puppeteer.launch({
        headless: false, // Set to false for visible browser windows
        args: [
          '--start-maximized',
          '--mute-audio',
          '--no-sandbox',
          '--disable-dev-shm-usage',
          `--user-agent=${user_agent}`,
        ],
      });

    const chromePath = 'chromedriver.exe';
    
    // const service = new chrome.ServiceBuilder().build();
    // const service = new chrome.ServiceBuilder(chromePath).build();
    // console.log(chrome);
    // chrome.setDefaultService(service);

    const page = await browser.newPage();
    await page.setViewport({ width: 1920, height: 1080 }); // Set viewport size

    //   the captcha solver should be initialized before visiting the page
    const captchaSolver = new CaptchaSolver(page)

    captchaSolver.solve({
        numAttempts: 5, // max number of attempts to solve the captcha
        startPosition: 25, // start position of the slider handle
        positionIncrement: 5, // number of pixels to advance the slider on each iteration
      })

    const username = account.username;
    const password = account.password;
    const proxy = proxies[Math.floor(Math.random() * proxies.length)];
    const comment = comments[Math.floor(Math.random() * comments.length)];

    if (logged_in_accounts.has(username)) {
      console.log(`Account ${username} is already logged in. Skipping...`);
      await browser.close();
      continue;
    }

    // // Set up Chrome options
    // const chromeOptions = {
    //   'args': [
    //     '--mute-audio',
    //     `--proxy-server=${proxy}`,
    //     '--no-sandbox',
    //     '--disable-dev-shm-usage',
    //     `--user-agent=${user_agent}`,
    //   ]
    // };

    // // Create a new WebDriver instance
    // const driver = await new Builder()
    //   .forBrowser('chrome')
    //   .setChromeOptions(chromeOptions)
    //   .build();

    try {
      console.log(`Account ${username} is logging in...`);

      await page.goto('https://www.tiktok.com/login/phone-or-email/email');

      await page.waitForSelector('input[name="username"]', { timeout: 10000 });
      await page.type('input[name="username"]', username);

      await page.waitForSelector('input[type="password"]', { timeout: 10000 });
      await page.type('input[type="password"]', password);

      await page.waitForTimeout(5000);

      // Press Enter to submit the login form
      await page.keyboard.press('Enter');

      // solve the captcha
      // await page.waitForXPath('/html/body/div[9]/div');
      await captchaSolver.solve()

      await page.waitForNavigation({ url: 'https://www.tiktok.com' });

      // Example: Check if login was successful
      if (page.url().includes('login')) {
        throw new Error('Login failed');
      } else {
        // Navigate to the target URL
        await page.goto(targetUrl);
      }

      // Add the account to the set of logged-in accounts
      logged_in_accounts.add(username);
       // Bot for like and comment
       await page.waitForXPath('/html/body/div[1]/div[2]/div[2]/div/div[2]/div[1]/div[1]/div[1]/div[4]/div/button[1]');
       const [likeButton] = await page.$x('/html/body/div[1]/div[2]/div[2]/div/div[2]/div[1]/div[1]/div[1]/div[4]/div/button[1]');
       const [tagElement] = await page.$x('/html/body/div[1]/div[3]/div[2]/div/div[2]/div[1]/div[1]/div[1]/div[4]/div/button[1]/span/div/div/svg/g/g[2]/g/g[4]');
 
      // Get the value of the 'style' attribute
      const styleValue = await tagElement.evaluate((element) => element.getAttribute('style'));

      // Perform a conditional action based on the style value
      if (!styleValue.includes('display: none;')) {
        await likeButton.click();
        await page.waitForTimeout(5000);

        // Add comment to the post
        const [commentClick] = await page.$x('/html/body/div[1]/div[2]/div[2]/div/div[2]/div[1]/div[3]/div[1]/div/div/div[1]/div');
        // const actions = driver.actions({ bridge: true });
        // await actions.move({ origin: commentClick }).perform();
        // await commentClick.click();

        // const commentInput = await driver.wait(until.elementLocated(By.xpath('/html/body/div[1]/div[2]/div[2]/div/div[2]/div[1]/div[3]/div[1]/div/div/div[1]/div/div[1]/div/div/div[2]/div/div/div/div/span')));
        // await commentInput.sendKeys(comment);
        // await driver.sleep(2000);
        // await commentInput.sendKeys(Key.ENTER);
        // await driver.sleep(2000);
      }

      // Logout and clear cookies for the next account
      // Delete all cookies
      const cookies = await page.cookies();
      await page.deleteCookie(...cookies);
    } catch (error) {
      console.log(`Error during login: ${error.message}`);
    } finally {
      await browser.close();
    }
  }
}

async function main() {
  const workers = 5; // multi-instance

  // Specify the file path of the Excel file
  const excelFile = 'AKUN GMAIL.xlsx';

  // Path to the external text file containing comment texts
  const commentsFile = 'comments.txt';

  // Path to the text file containing proxy addresses
  const proxyFilePath = 'proxy.txt';

  // URL to navigate to after logging in
  const targetUrl = 'https://www.tiktok.com/@nikxphreaker/video/6993669592551836930';

  // Read account credentials from the Excel file
  const accounts = [];
  const sheetName = 'MARSHA'; // Specify the sheet name if it's not the default sheet

  // Specify the columns and rows you want to extract
  const columns = ['Gmail', 'Password.2']; // Replace with your desired column names
  const startRow = 50; // Replace with the starting row index
  const endRow = 54; // Replace with the ending row index

  // Read the Excel file
  const XLSX = require('xlsx');
  const workbook = XLSX.readFile(excelFile);
  const worksheet = workbook.Sheets[sheetName];

  const data = XLSX.utils.sheet_to_json(worksheet, { header: 1 });

  // Extract the desired data range
  const selectedData = data.slice(startRow - 1, endRow);

  // Find the column indices based on column names
  // const column1Index = selectedData.indexOf(columns[0]);
  // const column2Index = selectedData.indexOf(columns[1]);

  // Extract data from the specified columns
  // const dataArray = selectedData.map(row => {
  //   if (row[1] != undefined || row[10] != undefined) {
  //     continue; // Skip rows with invalid data
  //   } else {
  //     [row[1], row[10]]
  //   }
    
  // });

  // console.log(selectedData);
  // console.log(dataArray);

  // // Convert the selected data into a list of objects
  for (const row of selectedData) {
    const username = row[1];
    const password = row[10];

    if (username == undefined || password == undefined) {
      continue; // Skip rows with invalid data
    }

    const account = { 'username' : username, 'password' : password };
    console.log(account);
    accounts.push(account);
  }

  // Read proxy addresses from the text file
  const fs = require('fs');
  const proxies = fs.readFileSync(proxyFilePath, 'utf-8').split('\n').map(line => line.trim()).filter(line => line !== '');

  // Read comment texts from the text file
  const comments = fs.readFileSync(commentsFile, 'utf-8').split('\n').map(line => line.trim()).filter(line => line !== '');

  const numPieces = 5; // Number of pieces to split the array into
  const arrayLength = accounts.length;
  const pieceSize = Math.floor(arrayLength / numPieces);
  const remainder = arrayLength % numPieces;

  const splitArrays = [];
  let start = 0;
  for (let i = 0; i < numPieces; i++) {
    const end = start + pieceSize + (i < remainder ? 1 : 0); // Adjust piece size for the remainder
    splitArrays.push(accounts.slice(start, end));
    start = end;
  }

  //Create an array of promises for each WebDriver instance
  const promises = splitArrays.map((accounts) => loginTikTok(accounts, proxies, targetUrl, comments));

  // Execute all promises concurrently
  await Promise.all(promises);

  console.log('All Selenium instances have completed.');
}

main().catch(console.error);
